Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm

```diff
@@ -301,6 +302,45 @@
     return dictionaryPopupInfo;
 }
 
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+void WebPage::getTextIndicatorForID(const WTF::UUID uuid, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&& completionHander)
+{
+    RefPtr range = m_unifiedTextReplacementController->rangaAssociatedWithUUID(uuid);
+    auto simpleRange = makeSimpleRange(range.get());
+
+    if (!simpleRange) {
+        completionHander(std::nullopt);
+        return;
+    }
+
+    if (!m_page) {
+        completionHander(std::nullopt);
+        return;
+    }
+
+    if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(m_page->mainFrame())) {
+        std::optional<TextIndicatorData> textIndicatorData;
+        constexpr OptionSet<TextIndicatorOption> textIndicatorOptions {
+            TextIndicatorOption::IncludeSnapshotOfAllVisibleContentWithoutSelection,
+            TextIndicatorOption::ExpandClipBeyondVisibleRect,
+            TextIndicatorOption::UseSelectionRectForSizing
+        };
+        if (auto textIndicator = TextIndicator::createWithRange(*simpleRange, textIndicatorOptions, TextIndicatorPresentationTransition::None, { }))
+            textIndicatorData = textIndicator->data();
+
+        completionHander(WTFMove(textIndicatorData));
+    }
```

Comment:
Make sure you call the `completionHandler` in the case where there's no `localMainFrame` as well.